kairos-prover/kairos-circuit-logic/src/account_trie.rs

```diff
@@ -0,0 +1,376 @@
+            pubkey,
+            balance,
+            nonce,
+        }
+    }
+
+    pub fn check_nonce(&self, nonce: u64) -> Result<(), TxnErr> {
+        if self.nonce != nonce {
+            return Err(format!(
+                "nonce mismatch: transaction nonce {nonce} does not match account nonce {}",
+                self.nonce,
+            ));
+        }
+
+        Ok(())
+    }
+
+    pub fn increment_nonce(&mut self) {
+        // TODO this is not a real problem an account will never have 2^64 transactions
+        // To make nonce wrapping safe
+        // we should expire transactions after a certain number of batches.
+        self.nonce = self.nonce.wrapping_add(1);
+    }
+}
+
+impl PortableHash for Account {
+    fn portable_hash<H: PortableUpdate>(&self, hasher: &mut H) {
+        self.pubkey.portable_hash(hasher);
+        self.balance.portable_hash(hasher);
+    }
+}
+
+/// A utility function to hash multiple buffers reusing the same hasher.
+fn hash_buffers<const N: usize, T: AsRef<[u8]>>(items: [T; N]) -> [KeyHash; N] {
+    let mut out = [KeyHash([0; 8]); N];
+    let mut hasher = Sha256::new();
+
+    for (i, item) in items.iter().enumerate() {
+        hasher.update(item.as_ref());
+        out[i] = KeyHash::from_bytes(&hasher.finalize_fixed_reset().into());
+    }
+
+    out
+}
+
+#[cfg(any(test, feature = "test-logic"))]
+pub mod test_logic {
+    use crate::{ProofInputs, ProofOutputs};
+
+    use super::*;
+    use alloc::rc::Rc;
+    use kairos_trie::{stored::memory_db::MemoryDb, DigestHasher};
+
+    pub fn test_prove_batch(
+        mut prior_root_hash: TrieRoot<NodeHash>,
+        db: Rc<MemoryDb<Account>>,
+        batches: Vec<Vec<KairosTransaction>>,
+        proving_hook: impl Fn(ProofInputs) -> Result<ProofOutputs, String>,
+    ) {
+        for batch in batches.into_iter() {
+            let mut account_trie = AccountTrie::new_try_from_db(db.clone(), prior_root_hash)
+                .expect("Failed to create account trie");
+            account_trie
+                .apply_batch(batch.iter().cloned())
+                .expect("Failed to apply batch");
+
+            let new_root_hash = account_trie
+                .txn
+                .commit(&mut DigestHasher::<sha2::Sha256>::default())
+                .expect("Failed to commit transaction");
+
+            let trie_snapshot = account_trie.txn.build_initial_snapshot();
+
+            let proof_inputs = ProofInputs {
+                transactions: batch.into_boxed_slice(),
+                trie_snapshot,
+            };
+
+            let ProofOutputs {
+                pre_batch_trie_root,
+                post_batch_trie_root,
+                deposits: _,
+                withdrawals: _,
+            } = proving_hook(proof_inputs).expect("Failed to prove execution");
+
+            assert_eq!(pre_batch_trie_root, prior_root_hash);
+            assert_eq!(post_batch_trie_root, new_root_hash);
+            prior_root_hash = new_root_hash;
```

Comment:
Can we return this instead of mutating it?

Reply:
no we have to provide the new root to the next batch that run in the next iteration of the loop.